group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupMetadataManager.java

```diff
@@ -848,26 +844,32 @@ ConsumerGroup consumerGroup(
     ConsumerGroup getOrMaybeCreatePersistedConsumerGroup(
         String groupId,
         boolean createIfNotExists
-    ) throws GroupIdNotFoundException {
+    ) throws IllegalStateException {
         Group group = groups.get(groupId);
 
         if (group == null && !createIfNotExists) {
-            throw new IllegalStateException(String.format("Consumer group %s not found.", groupId));
+            throw new IllegalStateException(String.format("Consumer group %s not found", groupId));
         }
 
         if (group == null) {
             ConsumerGroup consumerGroup = new ConsumerGroup(snapshotRegistry, groupId, metrics);
             groups.put(groupId, consumerGroup);
             metrics.onConsumerGroupStateTransition(null, consumerGroup.state());
             return consumerGroup;
+        } else if (group.type() == CONSUMER) {
+            return (ConsumerGroup) group;
+        } else if (group.type() == CLASSIC && ((ClassicGroup) group).isSimpleGroup()) {
+            // If the group is a simple classic group, it was automatically created to hold committed
+            // offsets if no group existed. Simple classic groups are not backed by any records
+            // in the __consumer_offsets topic hence we can safely replace it here. Without this,
+            // replaying consumer group records after offset commit records would not work.
```

Comment:
The issue happens when the offset commit records are before the consumer group records in the log (e.g. after compaction). When replaying, the following happens:
* OOM#replay will create a simple classic group if the group does not exist.
* GMM#replay for consumer group records fails because a classic group always exists.

This change handle the case in GMM#replay to basically swap the group if the group was a simple classic group.

Reply:
could you provide an example when we would GMM#replay consumer group records for a simple group? i don't think i understand how
> GMM#replay for consumer group records

can/will be triggered

Reply:
i guess `testCoordinatorFailoverCompactingPartitionWithManualOffsetCommitsAndConsumerGroupMemberUnsubscribingAndResubscribing` explains the example

Reply:
Just for my understanding -- these issues are all stemming from that when we designed the replay logic, we assumed the records would be as when they were first written, but since we compact, we may remove records (turn to tombstones case 1) or we may remove records that create groups and this was handled badly in cases 2 and 3

Reply:
Is 3 also a bug in how we handle the manual -> consumer group case?

Reply:
> could you provide an example when we would GMM#replay consumer group records for a simple group? i don't think i understand how

It could happen when you manually commit offsets and then form a consumer group. In this case, you have offset commit records followed by consumer group records.

Reply:
> Just for my understanding -- these issues are all stemming from that when we designed the replay logic, we assumed the records would be as when they were first written, but since we compact, we may remove records (turn to tombstones case 1) or we may remove records that create groups and this was handled badly in cases 2 and 3

Right.

Reply:
> Is 3 also a bug in how we handle the manual -> consumer group case?

I am not sure to understand the question. Manual commits uses a simple classic group so this new logic also covers it.

Reply:
Sorry if it was unclear -- the comment was saying that there was a bug in the compaction logic AND the logic for manual groups. (In other words, there were 2 bugs)